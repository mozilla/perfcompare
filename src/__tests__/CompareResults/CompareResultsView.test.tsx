import { renderHook } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { act } from 'react-dom/test-utils';

import CompareResultsTable from '../../components/CompareResults/CompareResultsTable';
import CompareResultsView from '../../components/CompareResults/CompareResultsView';
import SelectedRevisionsTable from '../../components/Shared/SelectedRevisionsTable';
import { setCompareResults } from '../../reducers/CompareResultsSlice';
import { updateSearchResults } from '../../reducers/SearchSlice';
import { setSelectedRevisions } from '../../reducers/SelectedRevisions';
import useProtocolTheme from '../../theme/ProtocolTheme';
import getTestData from '../utils/fixtures';
import { renderWithRouter, store } from '../utils/setupTests';
import { screen } from '../utils/test-utils';

describe('CompareResults View', () => {
  const protocolTheme = renderHook(() => useProtocolTheme()).result.current
    .protocolTheme;
  const themeMode = protocolTheme.palette.mode;

  it('Should match snapshot', () => {
    const { testCompareData } = getTestData();
    // set results data
    store.dispatch(setCompareResults(testCompareData));

    renderWithRouter(<CompareResultsTable theme={themeMode} />);
    expect(document.body).toMatchSnapshot();
  });

  it('should display SelectedRevisionsTable if there are selected revisions', async () => {
    const { testData } = getTestData();

    // start with four selected revisions
    const selectedRevisions = testData.slice(0, 4);
    store.dispatch(setSelectedRevisions(selectedRevisions));

    renderWithRouter(<CompareResultsView theme={protocolTheme} />);

    expect(screen.getByText("you've got no arms left!")).toBeInTheDocument();
  });

  it('should display revision search when clicking edit button', async () => {
    const { testData } = getTestData();
    // set delay to null to prevent test time-out due to useFakeTimers
    const user = userEvent.setup({ delay: null });

    // start with four selected revisions
    const selectedRevisions = testData.slice(0, 4);
    store.dispatch(setSelectedRevisions(selectedRevisions));

    renderWithRouter(<CompareResultsView theme={protocolTheme} />);

    const editButton = screen.getByRole('button', {
      name: 'edit-revision-1',
    });

    await user.click(editButton);

    act(() => void jest.runOnlyPendingTimers());
    expect(
      screen.getByRole('textbox', {
        name: 'Search By Revision ID or Author Email',
      }),
    ).toBeInTheDocument();
  });

  it('should show correct icon for improvements and regressions', () => {
    const { testCompareData } = getTestData();
    // set results data
    store.dispatch(setCompareResults(testCompareData));

    renderWithRouter(<CompareResultsTable theme={themeMode} />);

    expect(screen.getByTestId('ThumbUpAltIcon')).toBeInTheDocument();
    expect(screen.getByTestId('WarningIcon')).toBeInTheDocument();
  });
});

describe('SelectedRevisionsTableRow', () => {
  const protocolTheme = renderHook(() => useProtocolTheme()).result.current
    .protocolTheme;

  it('should close popover on close', async () => {
    const { testData } = getTestData();
    // set delay to null to prevent test time-out due to useFakeTimers
    const user = userEvent.setup({ delay: null });

    // start with four selected revisions
    const selectedRevisions = testData.slice(0, 4);
    store.dispatch(setSelectedRevisions(selectedRevisions));
    store.dispatch(updateSearchResults(testData));

    renderWithRouter(<SelectedRevisionsTable view='compare-results' />);

    await user.click(screen.getByRole('button', { name: 'edit-revision-1' }));

    const input = screen.getByRole('textbox', {
      name: 'Search By Revision ID or Author Email',
    });
    await user.click(input);

    expect(screen.getAllByText('spamspamspam')[0]).toBeInTheDocument();

    await user.keyboard('{Esc}');
    act(() => void jest.runOnlyPendingTimers());

    expect(screen.queryByText('It got better...')).not.toBeInTheDocument();
  });

  it('should close popover when close button is clicked', async () => {
    const { testData } = getTestData();
    store.dispatch(updateSearchResults(testData));
    // set delay to null to prevent test time-out due to useFakeTimers
    const user = userEvent.setup({ delay: null });

    // start with one selected revision
    const selectedRevisions = testData.slice(0, 1);
    store.dispatch(setSelectedRevisions(selectedRevisions));

    renderWithRouter(<CompareResultsView theme={protocolTheme} />);
    const editRevisionButton = screen.getByRole('button', {
      name: 'edit-revision-1',
    });
    expect(editRevisionButton).toBeInTheDocument();

    await user.click(editRevisionButton);

    const searchInput = screen.getByRole('textbox', {
      name: 'Search By Revision ID or Author Email',
    });
    expect(searchInput).toBeInTheDocument();

    await user.click(screen.getByTestId('cancel-edit-revision-button'));
    act(() => void jest.runOnlyPendingTimers());

    expect(searchInput).not.toBeInTheDocument();
  });
});

describe('CompareResultsTable', () => {
  const protocolTheme = renderHook(() => useProtocolTheme()).result.current
    .protocolTheme;
  const themeMode = protocolTheme.palette.mode;

  it('Should display correct confidence icon', async () => {
    const { testCompareData } = getTestData();
    // set results data
    store.dispatch(setCompareResults(testCompareData));

    renderWithRouter(<CompareResultsTable theme={themeMode} />);

    const icons = screen.getAllByTestId('confidence-icon');
    expect(icons[0]).toHaveClass('low');
    expect(icons[1]).toHaveClass('med');
    expect(icons[2]).toHaveClass('high');
    expect(icons[3]).toHaveClass('unknown-confidence');
  });
});
