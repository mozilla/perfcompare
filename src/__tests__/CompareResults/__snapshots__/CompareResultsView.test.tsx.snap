// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CompareResults View Should match snapshot 1`] = `
<body>
  <div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-w09ozj-MuiPaper-root-MuiTableContainer-root"
    >
      <table
        aria-label="a dense table"
        class="MuiTable-root css-d8sagy-MuiTable-root"
      >
        <thead
          class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-head css-1eza0l0-MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeSmall css-h76f5-MuiTableCell-root"
              scope="col"
            >
              Platform
              <button
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-15u9hm3-MuiButtonBase-root-MuiIconButton-root"
                data-testid="platform-options-button"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-fuz6rv-MuiSvgIcon-root"
                  data-testid="MoreVertIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
                  />
                </svg>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeSmall css-h76f5-MuiTableCell-root"
              scope="col"
            >
              Graph
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeSmall css-e3qqfc-MuiTableCell-root"
              scope="col"
            >
              Test Name
              <button
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-15u9hm3-MuiButtonBase-root-MuiIconButton-root"
                data-testid="test-options-button"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-fuz6rv-MuiSvgIcon-root"
                  data-testid="MoreVertIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
                  />
                </svg>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeSmall css-h76f5-MuiTableCell-root"
              scope="col"
            >
              Base
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeSmall css-h76f5-MuiTableCell-root"
              scope="col"
            >
              New
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeSmall css-h76f5-MuiTableCell-root"
              scope="col"
            >
              Delta
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeSmall css-h76f5-MuiTableCell-root"
              scope="col"
            >
              Status
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeSmall css-h76f5-MuiTableCell-root"
              scope="col"
            >
              Confidence
              <button
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-15u9hm3-MuiButtonBase-root-MuiIconButton-root"
                data-testid="confidence-options-button"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-fuz6rv-MuiSvgIcon-root"
                  data-testid="MoreVertIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
                  />
                </svg>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeSmall css-h76f5-MuiTableCell-root"
              scope="col"
            >
              Total Runs
            </th>
          </tr>
        </thead>
        <tbody
          class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-hover css-1eza0l0-MuiTableRow-root"
            data-testid="table-row"
          >
            <td
              aria-label="macosx1015-64-shippable-qr"
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall background-icon light-mode osx css-183p4wl-MuiTableCell-root"
              data-mui-internal-clone-element="true"
            />
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-183p4wl-MuiTableCell-root"
            >
              <a
                aria-label="Graph link"
                class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways background-icon light-mode graph-icon-color css-e4zel8-MuiTypography-root-MuiLink-root"
                href="https://treeherder.mozilla.org/perfherder/graphs?highlightedRevisions=1e32960788ea&highlightedRevisions=6a1f133815a1&series=mozilla-central%2C9daea6258ae244039eb2390f3eab7935994cbb0e%2C1%2C1&timerange=5184000"
                rel="noopener"
                target="_blank"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-fuz6rv-MuiSvgIcon-root"
                  data-testid="TimelineIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M23 8c0 1.1-.9 2-2 2-.18 0-.35-.02-.51-.07l-3.56 3.55c.05.16.07.34.07.52 0 1.1-.9 2-2 2s-2-.9-2-2c0-.18.02-.36.07-.52l-2.55-2.55c-.16.05-.34.07-.52.07s-.36-.02-.52-.07l-4.55 4.56c.05.16.07.33.07.51 0 1.1-.9 2-2 2s-2-.9-2-2 .9-2 2-2c.18 0 .35.02.51.07l4.56-4.55C8.02 9.36 8 9.18 8 9c0-1.1.9-2 2-2s2 .9 2 2c0 .18-.02.36-.07.52l2.55 2.55c.16-.05.34-.07.52-.07s.36.02.52.07l3.55-3.56C19.02 8.35 19 8.18 19 8c0-1.1.9-2 2-2s2 .9 2 2z"
                  />
                </svg>
              </a>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1cl4lun-MuiTableCell-root"
            >
              a11yr dhtml.html opt e10s fission stylo webrender
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-183p4wl-MuiTableCell-root"
            >
              704.84
               
              ms
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-183p4wl-MuiTableCell-root"
            >
              712.44
               
              ms
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-183p4wl-MuiTableCell-root"
            >
              1.08
              %
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-183p4wl-MuiTableCell-root"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-colorSuccess MuiSvgIcon-fontSizeMedium css-116ao24-MuiSvgIcon-root"
                data-testid="ThumbUpAltIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M2 20h2c.55 0 1-.45 1-1v-9c0-.55-.45-1-1-1H2v11zm19.83-7.12c.11-.25.17-.52.17-.8V11c0-1.1-.9-2-2-2h-5.5l.92-4.65c.05-.22.02-.46-.08-.66-.23-.45-.52-.86-.88-1.22L14 2 7.59 8.41C7.21 8.79 7 9.3 7 9.83v7.84C7 18.95 8.05 20 9.34 20h8.11c.7 0 1.36-.37 1.72-.97l2.66-6.15z"
                />
              </svg>
               
               
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall background-icon low css-183p4wl-MuiTableCell-root"
              data-testid="confidence-icon"
            />
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-183p4wl-MuiTableCell-root"
            >
              1
              /
              1
            </td>
          </tr>
          <tr
            class="MuiTableRow-root MuiTableRow-hover css-1eza0l0-MuiTableRow-root"
            data-testid="table-row"
          >
            <td
              aria-label="linux1804-64-shippable-qr"
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall background-icon light-mode linux css-183p4wl-MuiTableCell-root"
              data-mui-internal-clone-element="true"
            />
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-183p4wl-MuiTableCell-root"
            >
              <a
                aria-label="Graph link"
                class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways background-icon light-mode graph-icon-color css-e4zel8-MuiTypography-root-MuiLink-root"
                href="https://treeherder.mozilla.org/perfherder/graphs?highlightedRevisions=1e32960788ea&highlightedRevisions=6a1f133815a1&series=mozilla-central%2Cedcc6311f15bdf7924ef3f0ccfd8b47ce1892212%2C1%2C1&timerange=5184000"
                rel="noopener"
                target="_blank"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-fuz6rv-MuiSvgIcon-root"
                  data-testid="TimelineIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M23 8c0 1.1-.9 2-2 2-.18 0-.35-.02-.51-.07l-3.56 3.55c.05.16.07.34.07.52 0 1.1-.9 2-2 2s-2-.9-2-2c0-.18.02-.36.07-.52l-2.55-2.55c-.16.05-.34.07-.52.07s-.36-.02-.52-.07l-4.55 4.56c.05.16.07.33.07.51 0 1.1-.9 2-2 2s-2-.9-2-2 .9-2 2-2c.18 0 .35.02.51.07l4.56-4.55C8.02 9.36 8 9.18 8 9c0-1.1.9-2 2-2s2 .9 2 2c0 .18-.02.36-.07.52l2.55 2.55c.16-.05.34-.07.52-.07s.36.02.52.07l3.55-3.56C19.02 8.35 19 8.18 19 8c0-1.1.9-2 2-2s2 .9 2 2z"
                  />
                </svg>
              </a>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-183p4wl-MuiTableCell-root"
            >
              a11yr dhtml.html opt e10s fission stylo webrender
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-183p4wl-MuiTableCell-root"
            >
              776.97
               
              ms
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-183p4wl-MuiTableCell-root"
            >
              791.34
               
              ms
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-183p4wl-MuiTableCell-root"
            >
              1.85
              %
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-183p4wl-MuiTableCell-root"
            >
               
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-colorError MuiSvgIcon-fontSizeMedium css-1uz6t4g-MuiSvgIcon-root"
                data-testid="WarningIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M1 21h22L12 2 1 21zm12-3h-2v-2h2v2zm0-4h-2v-4h2v4z"
                />
              </svg>
               
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall background-icon med css-183p4wl-MuiTableCell-root"
              data-testid="confidence-icon"
            />
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-183p4wl-MuiTableCell-root"
            >
              1
              /
              1
            </td>
          </tr>
          <tr
            class="MuiTableRow-root MuiTableRow-hover css-1eza0l0-MuiTableRow-root"
            data-testid="table-row"
          >
            <td
              aria-label="windows10-64-shippable-qr"
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall background-icon light-mode windows css-1cl4lun-MuiTableCell-root"
              data-mui-internal-clone-element="true"
            />
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-183p4wl-MuiTableCell-root"
            >
              <a
                aria-label="Graph link"
                class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways background-icon light-mode graph-icon-color css-e4zel8-MuiTypography-root-MuiLink-root"
                href="https://treeherder.mozilla.org/perfherder/graphs?highlightedRevisions=1e32960788ea&highlightedRevisions=6a1f133815a1&series=mozilla-central%2C3a6d4b6178678c3113c980b378bb83b2629926a0%2C1%2C1&timerange=5184000"
                rel="noopener"
                target="_blank"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-fuz6rv-MuiSvgIcon-root"
                  data-testid="TimelineIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M23 8c0 1.1-.9 2-2 2-.18 0-.35-.02-.51-.07l-3.56 3.55c.05.16.07.34.07.52 0 1.1-.9 2-2 2s-2-.9-2-2c0-.18.02-.36.07-.52l-2.55-2.55c-.16.05-.34.07-.52.07s-.36-.02-.52-.07l-4.55 4.56c.05.16.07.33.07.51 0 1.1-.9 2-2 2s-2-.9-2-2 .9-2 2-2c.18 0 .35.02.51.07l4.56-4.55C8.02 9.36 8 9.18 8 9c0-1.1.9-2 2-2s2 .9 2 2c0 .18-.02.36-.07.52l2.55 2.55c.16-.05.34-.07.52-.07s.36.02.52.07l3.55-3.56C19.02 8.35 19 8.18 19 8c0-1.1.9-2 2-2s2 .9 2 2z"
                  />
                </svg>
              </a>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-183p4wl-MuiTableCell-root"
            >
              a11yr dhtml.html opt e10s fission stylo webrender
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-183p4wl-MuiTableCell-root"
            >
              643.54
               
              ms
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-183p4wl-MuiTableCell-root"
            >
              628.09
               
              ms
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1cl4lun-MuiTableCell-root"
            >
              -2.4
              %
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-183p4wl-MuiTableCell-root"
            >
               
               
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall background-icon high css-183p4wl-MuiTableCell-root"
              data-testid="confidence-icon"
            />
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-183p4wl-MuiTableCell-root"
            >
              1
              /
              1
            </td>
          </tr>
          <tr
            class="MuiTableRow-root MuiTableRow-hover css-1eza0l0-MuiTableRow-root"
            data-testid="table-row"
          >
            <td
              aria-label="windows10-64-shippable-qr"
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall background-icon light-mode windows css-183p4wl-MuiTableCell-root"
              data-mui-internal-clone-element="true"
            />
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-183p4wl-MuiTableCell-root"
            >
              <a
                aria-label="Graph link"
                class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways background-icon light-mode graph-icon-color css-e4zel8-MuiTypography-root-MuiLink-root"
                href="https://treeherder.mozilla.org/perfherder/graphs?highlightedRevisions=1e32960788ea&highlightedRevisions=6a1f133815a1&series=mozilla-central%2C3a6d4b6178678c3113c980b378bb83b2629926a0%2C1%2C1&timerange=5184000"
                rel="noopener"
                target="_blank"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-fuz6rv-MuiSvgIcon-root"
                  data-testid="TimelineIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M23 8c0 1.1-.9 2-2 2-.18 0-.35-.02-.51-.07l-3.56 3.55c.05.16.07.34.07.52 0 1.1-.9 2-2 2s-2-.9-2-2c0-.18.02-.36.07-.52l-2.55-2.55c-.16.05-.34.07-.52.07s-.36-.02-.52-.07l-4.55 4.56c.05.16.07.33.07.51 0 1.1-.9 2-2 2s-2-.9-2-2 .9-2 2-2c.18 0 .35.02.51.07l4.56-4.55C8.02 9.36 8 9.18 8 9c0-1.1.9-2 2-2s2 .9 2 2c0 .18-.02.36-.07.52l2.55 2.55c.16-.05.34-.07.52-.07s.36.02.52.07l3.55-3.56C19.02 8.35 19 8.18 19 8c0-1.1.9-2 2-2s2 .9 2 2z"
                  />
                </svg>
              </a>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-183p4wl-MuiTableCell-root"
            >
              a11yr dhtml.html opt e10s fission stylo webrender
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-183p4wl-MuiTableCell-root"
            >
              643.54
               
              ms
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-183p4wl-MuiTableCell-root"
            >
              628.09
               
              ms
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-183p4wl-MuiTableCell-root"
            >
              -2.4
              %
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-183p4wl-MuiTableCell-root"
            >
               
               
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall unknown-confidence css-183p4wl-MuiTableCell-root"
              data-testid="confidence-icon"
            >
              <button
                aria-label="Confidence not available"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium missing-confidence-button css-183p4wl-MuiButtonBase-root-MuiIconButton-root"
                data-mui-internal-clone-element="true"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium missing-confidence-icon css-fuz6rv-MuiSvgIcon-root"
                  data-testid="QuestionMarkIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M11.07 12.85c.77-1.39 2.25-2.21 3.11-3.44.91-1.29.4-3.7-2.18-3.7-1.69 0-2.52 1.28-2.87 2.34L6.54 6.96C7.25 4.83 9.18 3 11.99 3c2.35 0 3.96 1.07 4.78 2.41.7 1.15 1.11 3.3.03 4.9-1.2 1.77-2.35 2.31-2.97 3.45-.25.46-.35.76-.35 2.24h-2.89c-.01-.78-.13-2.05.48-3.15zM14 20c0 1.1-.9 2-2 2s-2-.9-2-2 .9-2 2-2 2 .9 2 2z"
                  />
                </svg>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-183p4wl-MuiTableCell-root"
            >
              1
              /
              1
            </td>
          </tr>
        </tbody>
        <tfoot
          class="MuiTableFooter-root css-9cgc89-MuiTableFooter-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-footer css-1eza0l0-MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-footer MuiTableCell-sizeSmall MuiTablePagination-root css-7o4z70-MuiTableCell-root-MuiTablePagination-root"
              colspan="8"
            >
              <div
                class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular MuiTablePagination-toolbar css-78c6dr-MuiToolbar-root-MuiTablePagination-toolbar"
              >
                <div
                  class="MuiTablePagination-spacer css-1psng7p-MuiTablePagination-spacer"
                />
                <p
                  class="MuiTablePagination-selectLabel css-437z57-MuiTablePagination-selectLabel"
                  id=":r6:"
                >
                  Rows per page:
                </p>
                <div
                  class="MuiInputBase-root MuiInputBase-colorPrimary css-9s8qn8-MuiInputBase-root-MuiTablePagination-select"
                  variant="standard"
                >
                  <select
                    aria-label="rows per page"
                    class="MuiTablePagination-select MuiNativeSelect-select MuiNativeSelect-standard MuiInputBase-input css-m3l1ig-MuiNativeSelect-select-MuiInputBase-input"
                    id=":r5:"
                  >
                    <option
                      class="MuiTablePagination-menuItem"
                      value="25"
                    >
                      25
                    </option>
                    <option
                      class="MuiTablePagination-menuItem"
                      value="50"
                    >
                      50
                    </option>
                    <option
                      class="MuiTablePagination-menuItem"
                      value="100"
                    >
                      100
                    </option>
                    <option
                      class="MuiTablePagination-menuItem"
                      value="-1"
                    >
                      All
                    </option>
                  </select>
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiTablePagination-selectIcon MuiNativeSelect-icon MuiNativeSelect-iconStandard css-l5belf-MuiSvgIcon-root-MuiNativeSelect-icon"
                    data-testid="ArrowDropDownIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M7 10l5 5 5-5z"
                    />
                  </svg>
                </div>
                <p
                  class="MuiTablePagination-displayedRows css-19c58e0-MuiTablePagination-displayedRows"
                >
                  1–4 of 4
                </p>
                <div
                  class="MuiBox-root css-1zye22"
                >
                  <button
                    aria-label="first page"
                    class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeMedium css-15u9hm3-MuiButtonBase-root-MuiIconButton-root"
                    disabled=""
                    tabindex="-1"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-fuz6rv-MuiSvgIcon-root"
                      data-testid="FirstPageIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M18.41 16.59 13.82 12l4.59-4.59L17 6l-6 6 6 6zM6 6h2v12H6z"
                      />
                    </svg>
                  </button>
                  <button
                    aria-label="previous page"
                    class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeMedium css-15u9hm3-MuiButtonBase-root-MuiIconButton-root"
                    disabled=""
                    tabindex="-1"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-fuz6rv-MuiSvgIcon-root"
                      data-testid="KeyboardArrowLeftIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M15.41 16.59 10.83 12l4.58-4.59L14 6l-6 6 6 6 1.41-1.41z"
                      />
                    </svg>
                  </button>
                  <button
                    aria-label="next page"
                    class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeMedium css-15u9hm3-MuiButtonBase-root-MuiIconButton-root"
                    disabled=""
                    tabindex="-1"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-fuz6rv-MuiSvgIcon-root"
                      data-testid="KeyboardArrowRightIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M8.59 16.59 13.17 12 8.59 7.41 10 6l6 6-6 6-1.41-1.41z"
                      />
                    </svg>
                  </button>
                  <button
                    aria-label="last page"
                    class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeMedium css-15u9hm3-MuiButtonBase-root-MuiIconButton-root"
                    disabled=""
                    tabindex="-1"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-fuz6rv-MuiSvgIcon-root"
                      data-testid="LastPageIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M5.59 7.41 10.18 12l-4.59 4.59L7 18l6-6-6-6zM16 6h2v12h-2z"
                      />
                    </svg>
                  </button>
                </div>
              </div>
            </td>
          </tr>
        </tfoot>
      </table>
    </div>
  </div>
</body>
`;
